### PR TITLE
feat(planner): support union/union all 

### LINKED ISSUE #7034: Feature: support union/union all for databend
**Summary**
https://www.postgresql.org/docs/current/queries-union.html (We have supported `intersect` and `except` set operator.)

`Union` also is used in recursive CTE.

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support union distinct and union all

Fixes #7034

### CODE REVIEW

File: src/query/service/src/sql/executor/pipeline_builder.rs
Comment (zhang2014) [inline src/query/service/src/sql/executor/pipeline_builder.rs:509]: Not need this if we use async_channel::bounded(2)?

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:42]:
```suggestion
impl AsyncSink for UnionReceiveSink {

}
```

File: src/query/service/src/sql/executor/pipeline_builder.rs
Comment (xudong963) [reply src/query/service/src/sql/executor/pipeline_builder.rs]: Yes, fixed

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:48]: Move this to `consume`?

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:62]:
```suggestion
        self.sender.send(data_block);
```

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs
Comment (xudong963) [reply src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:48]: fixed

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:30]:
```suggestion
    sender: Option<Sender<DataBlock>>,
```

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:44]:
```suggestion
        self.sender.take();
```

File: src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs:94]:
```suggestion
if let Ok(receiver_result) = self.receiver.try_recv() {
    self.receiver_result = receiver_result;
    return Ok(Event::Sync);
}
```

File: src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs:72]: remove this branch.

File: src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs:96]:
```suggestion
            if !self.finished {
                return Ok(Event::Async);
            }

            self.output.finish();
```

File: src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs:127]:
```suggestion
    async fn async_process(&mut self) -> Result<()> {
        if !self.finished   {
            if let Ok(receiver_result) = self.receiver.recv().await {
                    self.receiver_result = Some(receiver_result);
                    return Ok(());
            }

            self.finished = true;
        }

        Ok(())
    }
```

**Comment by xudong963** [comment]
@zhang2014 I think your comments are similar to the change https://github.com/datafuselabs/databend/pull/7160/commits/31b090fb850259ba684579a7459d8af027a60125 ?

File: src/query/service/src/sql/executor/pipeline_builder.rs
Comment (zhang2014) [inline src/query/service/src/sql/executor/pipeline_builder.rs:511]:
```suggestion
        self.main_pipeline.add_transform
```

**Comment by zhang2014** [comment]
> @zhang2014 I think your comments are similar to the change [31b090f](https://github.com/datafuselabs/databend/commit/31b090fb850259ba684579a7459d8af027a60125) ? Make receiver receive all messages

We should not wait for all data to be ready before merge `union` left and right

**Comment by xudong963** [comment]
> We should not wait for all data to be ready before merge `union` left and right

I got you, thanks for your detailed explanation, learn much!

**Review by zhang2014 — APPROVED** [review]

File: src/query/service/src/sql/executor/physical_plan.rs
Comment (leiysky) [inline src/query/service/src/sql/executor/physical_plan.rs:285]: Better rename it to `UnionAll`.

Comment (xudong963) [reply src/query/service/src/sql/executor/physical_plan.rs]: done

**Review by sundy-li — APPROVED** [review]

**Comment by xudong963** [comment]
@soyeric128 Please add union/union all to https://databend.rs/doc/reference/sql/query-operators/operators-set when you have time

**Comment by soyeric128** [comment]
> @soyeric128 Please add union/union all to https://databend.rs/doc/reference/sql/query-operators/operators-set when you have time

Ok

**Comment by soyeric128** [comment]
Please review https://github.com/datafuselabs/databend/pull/7281

**Pull request merged**
